File: Python/opa.py
import pyaop

class SecurityAspect(pyaop.Aspect):

    @pyaop.after("function()")
    def authenticate(self, result, *args, **kwargs):
        # lógica de autenticação de usuários

    @pyaop.before("function()")
    def authorize(self, *args, **kwargs):
        # lógica de autorização de usuários

class MyService:

    @pyaop.aspect(SecurityAspect())
    def my_function(self):
        # lógica da função

File: C/opa.c
#include "aop.h"

struct SecurityAspect {
    AOP_OBJECT
    // lógica de autenticação de usuários
    void (*authenticate)();
    // lógica de autorização de usuários
    void (*authorize)();
};

void authenticate() {
    // lógica de autenticação de usuários
}

void authorize() {
    // lógica de autorização de usuários
}

void my_function() {
    // lógica da função
}

int main() {
    struct SecurityAspect securityAspect = {AOP_OBJECT_INITIALIZER(authenticate, authorize)};
    aop_weave((void*) my_function, (void*) &securityAspect);
    my_function();
    return 0;
}
